src/ssd/ssd_shadow.c

```diff
@@ -0,0 +1,269 @@
+
+#include <assert.h>
+#include "common/scene-helpers.h"
+#include "labwc.h"
+#include "buffer.h"
+#include "ssd-internal.h"
+#include "theme.h"
+#include "view.h"
+#include <cairo.h>
+
+#define FOR_EACH_STATE(ssd, tmp) FOR_EACH(tmp, \
+	&(ssd)->shadow.active, \
+	&(ssd)->shadow.inactive)
+
+/*
+ * Implements point_accepts_input for a buffer which never accepts input
+ * because drop-shadows should never catch clicks!
+ */
+static bool
+never_accepts_input(struct wlr_scene_buffer *buffer, double *sx, double *sy)
+{
+	return false;
+}
+
+static void
+set_shadow_geometry(struct ssd *ssd)
+{
+	struct view *view = ssd->view;
+	struct theme *theme = view->server->theme;
+	int titlebar_height = ssd->titlebar.height;
+	int width = view->current.width;
+	int height = view_effective_height(view, false) + titlebar_height;
+
+	struct ssd_part *part;
+	struct ssd_sub_tree *subtree;
+	struct wlr_fbox src_box;
+	int x;
+	int y;
+
+	FOR_EACH_STATE(ssd, subtree) {
+		bool active = subtree == &ssd->shadow.active;
+		int visible_shadow_width = active ? theme->window_active_shadow_radius
+			: theme->window_inactive_shadow_radius;
+		/* inset as a proportion of shadow width */
+		double inset_proportion = active ? theme->window_active_shadow_inset
+			: theme->window_inactive_shadow_inset;
+		/* inset in actual pixels */
+		int inset = inset_proportion * (double)visible_shadow_width;
+
+		/* If the shadow inset is greater than half the overall window
+		 * height (eg. becaused the window is shaded) then the corners
+		 * would overlap, which looks horrible.  To avoid this use
+		 * wlr_scene_buffer_set_source_box() to clip off the portion of
+		 * the corners which would overlap.  This does produce slight
+		 * aberrations in the shadow shape but it's not too noticeable.
+		 */
+		bool overlap = inset > (height / 2 - 1);
+		int overlap_px = overlap ? inset - height / 2 : 0;
+
+		/* Total size of corner buffers including inset and visible
+		 * portion.  Top and bottom are the same size (only the cutout
+		 * is different).  The buffers are square so width == height.
+		 */
+		int corner_size = active
+			? theme->shadow_corner_top_active->unscaled_height
+			: theme->shadow_corner_top_inactive->unscaled_height;
+
+		/* If titlebar height is odd and we're omitting the right/left
+		 * sides due to overlap, then we need to bring up the bottom
+		 * edge by one pixel to avoid leaving a gap.
+		 */
+		int avoid_gap = overlap && (titlebar_height % 2 == 1);
+
+		wl_list_for_each(part, &subtree->parts, link) {
+			struct wlr_scene_buffer *scene_buf =
+				wlr_scene_buffer_from_node(part->node);
+			switch (part->type) {
+			case LAB_SSD_PART_CORNER_BOTTOM_RIGHT:
+				x = width - inset;
+				y = -titlebar_height + height - inset
+					+ overlap_px - avoid_gap;
+				wlr_scene_node_set_position(part->node, x, y);
+
+				src_box.x = 0;
+				src_box.y = overlap_px;
+				src_box.width = corner_size;
+				src_box.height = corner_size - overlap_px;
```

Comment:
Can we somehow reduce code duplication here? There seems to be only one box with its axis changed on two occasions. Maybe we could define it once outside of the loop and have a static inline function to create a copy and switch `.x` and `.y` and `.width` and `.height` before returning it?

Or as alternative, what about something like this:
```c
void
buffer_transform(struct wlr_scene_buffer *buffer, int overlap_px,
		int corner_size, bool switch_axis)
{
	struct wlr_fbox box = {
		.x = switch_axis ? overlap_px : 0,
		.y = switch_axis ? 0 : overlap_px,
		.width = switch_axis ? (corner_size - overlap_pix) : corner_size,
		.height = switch_axis ? corner_size : (corner_size - overlap_pix)
	};
	wlr_scene_buffer_set_source_box(buffer, &box);
	wlr_scene_buffer_set_dest_size(buffer, box.width, box.height);
}
```